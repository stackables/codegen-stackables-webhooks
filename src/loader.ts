import { parse } from "graphql";
import { gql, GraphQLClient } from "graphql-request";

export default async function (
	endpoint: string,
	_: unknown,
	conf: Record<string, { bearerToken: string }>
) {
	const graphQLClient = new GraphQLClient(endpoint, {
		headers: {
			authorization: `Bearer ${conf[endpoint].bearerToken}`,
		},
	});

	const query = gql`
		query CodeGenWebhooks {
			Webhook: _Webhook {
				list {
					name
					fragment
					entity {
						id
					}
				}
			}
		}
	`;

	const data = await graphQLClient.request(query);
	const fragments = data.Webhook.list;

	return parse(
		fragments
			.map(
				(f: any) =>
					`fragment ${f.name} on ${f.entity.id} @register ${f.fragment}`
			)
			.join("\n")
	);
}
