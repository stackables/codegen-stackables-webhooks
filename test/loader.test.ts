import loader from '../src'

jest.setTimeout(30000)

test('Test Loader', async () => {
    const output = await loader('https://data.stackables.io/stackables-demo:prod/example1?introspection=a89751eb-ef0d-4663-9530-fa683b9d8483')
    expect(output).toMatchSnapshot()
})
