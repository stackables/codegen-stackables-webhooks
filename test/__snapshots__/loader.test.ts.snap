// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Loader 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [
        Object {
          "arguments": Array [],
          "kind": "Directive",
          "loc": Object {
            "end": 59,
            "start": 50,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 59,
              "start": 51,
            },
            "value": "register",
          },
        },
      ],
      "kind": "FragmentDefinition",
      "loc": Object {
        "end": 68,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 31,
          "start": 9,
        },
        "value": "CleanupAfterNodeDelete",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": Object {
          "end": 68,
          "start": 60,
        },
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "loc": Object {
              "end": 66,
              "start": 64,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 66,
                "start": 64,
              },
              "value": "id",
            },
            "selectionSet": undefined,
          },
        ],
      },
      "typeCondition": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 49,
          "start": 35,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 49,
            "start": 35,
          },
          "value": "DataEntityNode",
        },
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 68,
    "start": 0,
  },
}
`;
